### PR TITLE
Add new subscribe API

### PR DESCRIPTION
**Pull request by nirinchev**
## Description

Fixes RNET-865

Adds `IQueryable.SubscribeAsync` syntax sugar.
Adds `cancellationToken` argument to `Session.WaitForDownloadAsync`, `Session.WaitForUploadAsync`, and `SubscriptionSet.WaitForSynchronization`. This doesn't cancel the underlying Core wait operation, just the managed task. While it's not ideal, the native wait operation is not really expensive and it will eventually complete anyway, so we are not leaking resources.

There are some whitespace changes, so it might be a good idea to review with the "Ignore whitespace" option.

##  TODO

* [x] Changelog entry
* [x] Tests (if applicable)

### CODE REVIEW

File: Realm/Realm/Handles/SessionHandle.cs [hunk 1]
```diff
@@ -209,22 +210,36 @@ public void UnsubscribeNotifications()
             }
         }
 
-        public async Task WaitAsync(ProgressDirection direction)
+        public async Task WaitAsync(ProgressDirection direction, CancellationToken? cancellationToken)
         {
             var tcs = new TaskCompletionSource();
+            if (cancellationToken?.IsCancellationRequested == true)
+            {
+                tcs.TrySetCanceled(cancellationToken.Value);
```
Comment (papafe) [inline Realm/Realm/Handles/SessionHandle.cs:218]: I am not sure I understand why we need this and the following line. Can we just return?

**Review by papafe — APPROVED** [review]
Looks good. 
By the way, it was a little bit difficult to see the body of `SubscribeAsync` together  with the removal of the namespace block 😁

File: Tests/Realm.Tests/Database/APITests.cs [hunk 2]
```diff
@@ -19,6 +19,7 @@
 using System;
 using System.Linq;
 using System.Reflection;
+using System.Threading;
```
[the remark this answers is not in the document]
Comment (nirinchev) [reply Tests/Realm.Tests/Database/APITests.cs:22]: We don't - had some tests there, that I then removed and this was a leftover.

File: Realm/Realm/Handles/SessionHandle.cs (hunk 1 above)
Comment (nirinchev) [reply Realm/Realm/Handles/SessionHandle.cs:218]: If cancellation is requested, the expected behavior would be to throw a `TaskCancelledException`, even if we didn't start doing the work. Regarding why we're cancelling the tcs vs throwing a `TaskCancelledException`, I guess both are roughly the same and don't have strong feelings about one or the other.

**Comment by nirinchev** [comment]
> By the way, it was a little bit difficult to see the body of SubscribeAsync together with the removal of the namespace block

That's why you're supposed to read the PR description 😜

**Review by papafe — APPROVED** [review]

**Pull request merged**